### PR TITLE
Cmo api

### PR DESCRIPTION
**Pull request by chmorgan**
@DavidAntliff here is my latest code. I've also pushed changes to a forked version of the example project. This has the latest driver code integrated but I'll probably go back and rename .uninitialize to .close or something easier to type. Thoughts?

I'm not able to get the RMT input working for some reason. Testing here shows perfect output timing, minimal or no bus contention and the device responding, but there is no valid input reported from the RMT. I'm stumped and would appreciate another set of eyes.

### CODE REVIEW

**Comment by chmorgan** [comment]
@DavidAntliff had a chance to take a look at this yet? I also asked on the ESP32 forums for some help, had one suggestion but nothing that looks related to the rmt rx issue that I'm seeing with this code.

**Comment by DavidAntliff** [comment]
Not yet, I should have time later today or tomorrow. I'll see if anything about RMT RX stands out to me. Thank you for your efforts.

**Comment by chmorgan** [comment]
@DavidAntliff I may have one lead. I don't believe I'm enabling input on that gpio pin and going to try that once I wrap up a couple of other things.

**Comment by DavidAntliff** [comment]
Any luck with the GPIO input enable?

**Comment by chmorgan** [comment]
I should be back to try that today. After fixing a few things in libesphttpd I was looking into why the gpio driver has inconsistent measurement times and unfortunately can’t see any reason for it beyond other operations being done in the processor.

I’ll let you know as soon as I’ve retested it.

**Comment by chmorgan** [comment]
@DavidAntliff and enabling open drain fixed the issue. Everything looks solid, timing looks good, temperatures look good and no crc errors so far!

Thoughts on when you could take a look at it? I'd prefer working from your main repo rather than my fork.

**Comment by DavidAntliff** [comment]
Sounds good - I will look at it and test with my circuit tonight (~12 hours from now).

**Comment by DavidAntliff** [comment]
Am I right in thinking your PR needs the "latest" ESP-IDF rather than v2.1.1? I'm having trouble with `rmt_get_ringbuf_handle` - doesn't seem to be in 2.1.1.

EDIT: I think I worked it out - it was renamed in "latest" - needs to be `rmt_get_ringbuf_handler` in 2.1.

File: owb_rmt.c [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+            rmt_rx.mem_block_num = 1;
+            rmt_rx.rmt_mode = RMT_MODE_RX;
+            rmt_rx.rx_config.filter_en = true;
+            rmt_rx.rx_config.filter_ticks_thresh = 30;
+            rmt_rx.rx_config.idle_threshold = OW_DURATION_RX_IDLE;
+            if (rmt_config( &rmt_rx ) == ESP_OK)
+            {
+                if (rmt_driver_install( rmt_rx.channel, 512, ESP_INTR_FLAG_LOWMED | ESP_INTR_FLAG_IRAM | ESP_INTR_FLAG_SHARED ) == ESP_OK)
+                {
+                    rmt_get_ringbuf_handle( info->rx_channel, &info->rb );
```
Comment (DavidAntliff) [inline owb_rmt.c:384]: Can you add a #ifdef based on ESP-IDF version for this function? For version 2.1.1, it needs to be `rmt_get_ringbuf_handler` and changes to `rmt_get_ringbuf_handle` in the (unreleased) head of master. Not exactly sure how to deal with this since there is no release of ESP-IDF to date with the new name. If you need to use "latest" then we may need to rethink how we handle this. Thoughts?

**Review by DavidAntliff — COMMENTED** [review]
Chris, this is good - it works well with my three sensors. Thank you for your effort! Before I merge it, there are a few things we should address:

1. There's an API change for `rmt_get_ringbuf_handle` - see earlier comment. How best to handle this?
2. I'd like to tidy up the static/dynamic initialization but I can do that post-merge - it will change the API slightly though. Hopefully that won't be a problem.
3. There are a few old functions/macros brought in from the nodemcu code - e.g. `PIN_INPUT_ENABLE` is a direct macro I think and no longer officially part of the ESP-IDF API. It might be safer to use `gpio_set_direction`. However I can tidy these calls up post-merge.

I manually modified the esp32-ds18b20-example project to suit the new API (I saw your WIP PR but I wanted to change a few things) and I had to alter esp32-ds18b20 too slightly.

To summarise: it's great to see the RMT peripheral working (and therefore avoiding interrupt disabling), and I'm happy to merge this soon. I want to refine the API some more post-merge.

File: owb_rmt.c (hunk 1 above)
Comment (chmorgan) [reply owb_rmt.c:384]: @DavidAntliff 

I was hoping we could retain the new externally determined static/dynamic approach. I removed the allocation/free routines because the end user could simply malloc the structure if they'd like, or statically allocate it. It didn't seem necessary to duplicate functionality that is built into the language, it just made the library more complex. It confused me the first time I used it actually.

The PIN_INPUT_ENABLE exists in the latest esp-idf and is defined in a rom header include file in esp-idf. I found I had to use those routines directly as the gpio driver routines reconfigured the pins into gpio mode, disabling rmt mode.

I see there is an IDF_VER variable and a esp_get_idf_version() but I don't see an way to tell the version at compile time. We could always ask or document that you need latest to run with the RMT. I'm guessing 3.0 will release relatively soon.

Comment (DavidAntliff) [reply owb_rmt.c:384]: @chmorgan I hear you. I can see how the interface I had might be confusing. If it helps, this is what I'm trying to do:

https://github.com/DavidAntliff/esp32-components/wiki/Component-Design

In my experience with building large commercial apps with C, providing external allocation can cause problems with encapsulation and future API changes. I've used heavier techniques in other code bases so I'm trying to create a more lightweight solution that still makes some sense. I admit it's not perfect and I'm still trying to find a good compromise.

Fair point on the `PIN_INPUT_ENABLE` macro - if it works (and it does) then that's fine for now.

I'm not sure when 3.0 will be released - it's been "pending" for many months. However we could define a macro that can be easily changed (and commented) between API versions. Then the main line can be set for 2.x, and you can easily use a compiler directive to override it in your "pre-3.0" build.

Comment (chmorgan) [reply owb_rmt.c:384]: @DavidAntliff I can appreciate the consistent approach with the modules. I'm sure you noticed but to handle the instantiate/cleanup along with api and internal structure changes I created driver specific init function and a generic cleanup function. It's your repository and your choice on how you'd like it to look in the end!

Not sure when 3.0 will release but maybe a compile time define would work for that. Then users of the older or newer esp-idf (I'd probably assume the newer esp-idf by default but again your choice) can set the appropriate define in their build.

Comment (DavidAntliff) [reply owb_rmt.c:384]: Ok, in order to move forward, how about I merge this PR (and also merge the changes in esp32-ds18b20 and esp32-ds18b20-example I have made to match the API changes ), on the understanding that I'd like to revisit the API myself, so I make no guarantees that it might not change in the near or distant future? After that, if it's not quite how you like it, then you could either wrap it, or even fork if you wish. Sound OK? :)

I'll probably also add a compile-time define some time after the merge for selecting between the 2.x or future, but for now I'm happy to leave it building with "pre-3.0" since that's the future of ESP-IDF (at least as far as we know at this time).

**Comment by chmorgan** [comment]
@davidantliff sounds good to me! I’m really ok either way with the api but thought I’d explain why I did it. Oh, and from my work in embedded I always favor static allocation but the esp32 has so much memory it looks like people mostly don’t care! I’m used to the stm32 :-)

Did you see the changes I made to the ds18b20 library and example? I think I pushed those for your review but didn’t add pull requests to them. If you’d like me to take a look just let me know!

**Comment by chmorgan** [comment]
@davidantliff I also did a write up of the timing issues with the gpio driver. They look caused by overall system load. http://chmorgan.blogspot.com/2017/11/oscilloscope-assisted-debugging.html?m=1

**Comment by DavidAntliff** [comment]
Yes, I'm starting to find myself just dynamically allocate data structures on the ESP32 myself - plenty of RAM for my applications so far. It does make things a bit simpler, and if performance is a concern then I just allocate them early. However I like to keep the option for static allocation since that's also what I'm used to.

I saw your changes in the other libraries and I've pretty much copied them into my own branches, with a few editorial changes (I wanted to keep some of the old code commented out until I get a chance to revisit). You can submit PRs for your forks if you like, otherwise I'll just merge these:

https://github.com/DavidAntliff/esp32-ds18b20/commits/esp32-owb-pr4
https://github.com/DavidAntliff/esp32-ds18b20-example/commits/esp32-owb-pr4

Thanks also for the writeup. I've always felt that the GPIO approach was a compromise and I also ran into issues when I enabled WiFi. It was always my intention to use a peripheral like RMT (or similar) but I didn't find the time. Thankfully you've been able to contribute, and for that I appreciate your efforts.

**Comment by chmorgan** [comment]
I really wanted those crc errors fixed! Thanks for all of the help along the way.

**Comment by DavidAntliff** [comment]
@chmorgan Just for the record, I have had my ESP32 sampling from three DS18B20 devices every second for almost the last 10 days (+850k samples), using this new approach, and I have had zero CRC errors. Good work!

**Comment by chmorgan** [comment]
Very cool! I’d say that was a good test run.

**Comment by DavidAntliff** [comment]
@chmorgan sorry to bring this up again however I wonder if you are still using this code? I ask because I've been having some random crashes with the RMT driver with version 3.0 (didn't seem to happen in 2.1.1) and I'm not quite sure what's going on. Have you seen anything similar? Note that I'm using RMT with my multiple temp sensors (just one TX and RX RMT channel for all of them), but I'm also simultaneously using a different RMT channel with another type of sensor on a different GPIO, counting hall-effect pulses from a flow meter, so it might be an issue with the combination of sensors. So far I've traced the problem to a null pointer in the ESP-IDF code. I'll report it to the ESP-IDF project once I work out what is actually happening.

**Comment by chmorgan** [comment]
Hi Dave.

I haven't seen any crashes here in the rmt code. I'm using esp-idf latest
master as of yesterday. Only with one OneWire sensor though.

We do lean heavily on the rmt driver but of course its possible the bug is
in our code somewhere. I'd be interested in whether you find it.

I am seeing memory corruption in the heap here with some other code I'm
working on. It has been quite a challenge and I don't have it figured out
yet.

Chris

**Comment by DavidAntliff** [comment]
Ok, good to know that it's working OK for you in your case. That at least helps suggest that I may be running into an issue with _multiple_ sensors (or the fact that I'm creating multiple instances of the device on a single owb instance). What I see, when I enable the temp sensor code, is OK operation for some time (minutes) but then a random crash, such that when it starts crashing it then repeatedly crashes over and over at sensor initialisation. Only a _long_ reset press or power cycle recovers from this.

Unrelated perhaps - I've also noticed, since moving to the 3.0 branch, that the serial console (make monitor) is pretty unreliable. With 2.1.1 I could leave it open and connected for days, but with 3.0 it barely lasts an hour before getting a "Device not configured" error in the Python `read` call. However this happens even if the temp sensor code is disabled.

Have you been able to use any of the stack and heap corruption detection options in the newer ESP IDF? I've seen them in `make menuconfig` but haven't tried them yet.

**Comment by chmorgan** [comment]
I’ve been trying with the heap debugging features. I’m seeing the heap
corruption in the exact same location so I’m closer than the other day but
still not there yet. I keep reviewing code hoping I’ll spot it. This code
is doing a ton of realloc calls so if it’s an issue in another component
this is the code that would catch it due to the comprehensive heap checking
running during the realloc calls.
